pkg/controllermanager/controller/project/project_control_reconcile.go

```diff
@@ -252,6 +273,67 @@ func (c *defaultControl) reconcileNamespaceForProject(ctx context.Context, garde
 	return namespace, nil
 }
 
+// quotaConfiguration returns the first matching quota configuration if one is configured for the given project.
+func quotaConfiguration(config config.ControllerManagerControllerConfiguration, project *gardencorev1beta1.Project) (*config.QuotaConfiguration, error) {
+	if config.Project == nil {
+		return nil, nil
+	}
+
+	for _, c := range config.Project.Quotas {
+		quotaConfig := c
+		selector, err := metav1.LabelSelectorAsSelector(quotaConfig.ProjectSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		if selector.Matches(labels.Set(project.GetLabels())) {
+			return &quotaConfig, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// ResourceQuotaName is the name of the default ResourceQuota resource that is created by Gardener in the project namespace.
+const ResourceQuotaName = "gardener"
+
+func createOrUpdateResourceQuota(ctx context.Context, c client.Client, projectNamespace string, ownerReference *metav1.OwnerReference, config config.QuotaConfiguration) error {
+	resourceQuota, ok := config.Config.(*corev1.ResourceQuota)
+	if !ok {
+		return fmt.Errorf("failure while reading ResourceQuota from configuration: %v", resourceQuota)
+	}
+
+	projectResourceQuota := &corev1.ResourceQuota{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ResourceQuotaName,
+			Namespace: projectNamespace,
+		},
+	}
+	if err := c.Get(ctx, kutils.KeyFromObject(projectResourceQuota), projectResourceQuota); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	quotas := make(map[corev1.ResourceName]resource.Quantity)
+	for resourceName, quantity := range resourceQuota.Spec.Hard {
+		if val, ok := projectResourceQuota.Spec.Hard[resourceName]; ok {
+			// Do not overwrite already existing quotas.
+			quotas[resourceName] = val
+			continue
+		}
+		quotas[resourceName] = quantity
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, c, projectResourceQuota, func() error {
+		projectResourceQuota.SetOwnerReferences(common.MergeOwnerReferences(projectResourceQuota.GetOwnerReferences(), *ownerReference))
+		projectResourceQuota.Spec.Hard = quotas
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Can you avoid the extra `Get` call in `CreateOrUpdate`?
WDYT about using `Patch` here, if the quota doesn't exist, yet?

Reply:
Thanks @timebertt, definitely makes sense to remove the extra Get call. However, I'm sticking to `CreateOrUpdate` to keep things simple and since we are in the same cluster, I don't expect to many performance benefits. Do you agree?

Reply:
> I don't expect too many performance benefits with a `PATCH` request.

Yes, I wouldn't expect that either. As you moved the quota merging into the `CreateOrUpdate`, this is already better than before with the extra `Get` call. Perfect.
Thanks!